### PR TITLE
Add `on_finish_action` to `KubernetesPodOperator`

### PR DESCRIPTION
**Pull request by hussein-awala**
Add a new argument `on_finish_action` for `KubernetesPodOperator` to determine how to deal with the finished pods. There are 3 possible values:
- `delete_pod`: delete the pod regardless its state
- `delete_succeeded_pod`: delete the pod if it terminate with `Succeeded` state
- `keep_pod`: keep the pod and don't delete it

### CODE REVIEW

File: airflow/providers/cncf/kubernetes/operators/pod.py
Comment (eladkal) [inline airflow/providers/cncf/kubernetes/operators/pod.py]: This feels a bit complicated.
I think we should consider replace is_delete_operator with:
action_on_finish that supports several modes. For example: delete_pod, delete_pod_success

That way you dont need to work hard to understand the combinationa of flags, the mode is self explanatory.

WDYT?

Comment (hussein-awala) [reply airflow/providers/cncf/kubernetes/operators/pod.py]: My first thought was doing something similar to #29394, but I hesitated to deprecate `is_delete_operator_pod`, but yes, these flags may confuse users. I'll implement the argument you propose.

Comment (eladkal) [reply airflow/providers/cncf/kubernetes/operators/pod.py]: cc @jedcunningham WDYT?

Comment (potiuk) [reply airflow/providers/cncf/kubernetes/operators/pod.py]: Yep. Agree with @eladkal simple flag is better. I made similar "keep old flag" approaches in the past and it led to really bad complexity/

**Review by potiuk — APPROVED** [review]
I like this. Needs conflict resolving though. @jedcunningham @dstandish ?

File: airflow/providers/cncf/kubernetes/operators/pod.py
Comment (eladkal) [inline airflow/providers/cncf/kubernetes/operators/pod.py:230]: How do we feel about
`cleanup_action`?
with options `keep`, `delete`, `delete_succeeded` ?

Comment (potiuk) [reply airflow/providers/cncf/kubernetes/operators/pod.py:230]: `on_finish_action` is better IMHO -> `cleanup = keep` sounds redundant and contradicting (are we cleaning up really if we keep it ?)

File: airflow/providers/cncf/kubernetes/operators/pod.py
Comment (jedcunningham) [inline airflow/providers/cncf/kubernetes/operators/pod.py]:
```suggestion
                DeprecationWarning,
                stacklevel=2,
```

We probably should do this (think this is the right level).

File: tests/providers/cncf/kubernetes/operators/test_pod.py
Comment (jedcunningham) [inline tests/providers/cncf/kubernetes/operators/test_pod.py]:
```suggestion
            ),  # check backcompat of is_delete_operator_pod
```

This might be better, took me a bit to figure out what "b/c" was.

File: tests/providers/cncf/kubernetes/triggers/test_pod.py
Comment (jedcunningham) [inline tests/providers/cncf/kubernetes/triggers/test_pod.py]:
```suggestion
        Test that KubernetesPodTrigger fires the correct event if the task was cancelled.
```

nit

**Review by jedcunningham — COMMENTED** [review]
System test examples need updating, particularly as they get sucked into the docs.

Are you planning to change the operators based on KPO in a followup PR? Just want to make sure it isn't missed. For example, the [eks pod operator](https://github.com/apache/airflow/blob/4c5cebf04555aa7b892a50a85f2f488acc4c057b/airflow/providers/amazon/aws/operators/eks.py#L685-L697).

**Comment by hussein-awala** [comment]
> Are you planning to change the operators based on KPO in a followup PR? Just want to make sure it isn't missed. For example, the [eks pod operator](https://github.com/apache/airflow/blob/4c5cebf04555aa7b892a50a85f2f488acc4c057b/airflow/providers/amazon/aws/operators/eks.py#L685-L697).

@jedcunningham yes, in this PR, I have ensured that it is fully backward compatible. This means that users can update the `cncf.kubernetes` provider version and still use it with older versions of Amazon or Google providers:
https://github.com/apache/airflow/blob/a99e3ba37ee52a2a63998979d492213e0a68a24c/airflow/providers/cncf/kubernetes/operators/pod.py#L387-L393

Once this PR is merged, I'll create a separate PR for each provider to deprecate the argument `is_delete_operator_pod` and use the new one.

File: airflow/providers/cncf/kubernetes/operators/pod.py
Comment (hussein-awala) [reply airflow/providers/cncf/kubernetes/operators/pod.py:230]: I think the current action names are more clear. However, for the argument name, I have no preference between `cleanup_action` and `on_finish_action`; both are meaningful and suitable.

Comment (eladkal) [reply airflow/providers/cncf/kubernetes/operators/pod.py:230]: Cool lets stay with that

**Comment by eladkal** [comment]
> > Are you planning to change the operators based on KPO in a followup PR? Just want to make sure it isn't missed. For example, the [eks pod operator](https://github.com/apache/airflow/blob/4c5cebf04555aa7b892a50a85f2f488acc4c057b/airflow/providers/amazon/aws/operators/eks.py#L685-L697).
> 
> @jedcunningham yes, in this PR, I have ensured that it is fully backward compatible. This means that users can update the `cncf.kubernetes` provider version and still use it with older versions of Amazon or Google providers:
> 
> https://github.com/apache/airflow/blob/a99e3ba37ee52a2a63998979d492213e0a68a24c/airflow/providers/cncf/kubernetes/operators/pod.py#L387-L393
> 
> Once this PR is merged, I'll create a separate PR for each provider to deprecate the argument `is_delete_operator_pod` and use the new one.

We can do it in this PR if you prefer.. no reason to keep it in seperated PRs unless that is how you prefer.

**Comment by hussein-awala** [comment]
@eladkal we have this message added 15 months ago in the version 3.0.0 of Amazon provider, and we are releasing the version 8.0.0 without doing anything
> :param is_delete_operator_pod: What to do when the pod reaches its final
        state, or the execution is interrupted. If True, delete the
        pod; if False, leave the pod.  Current default is False, but this will be
        changed in the next major release of this provider.

Also we have a warning on the None `is_delete_operator_pod`, what is the policy for the providers deprecation?

**Comment by hussein-awala** [comment]
Same warning added in to Google provider in 6.4.0, and we are releasing 10.0.0

**Comment by potiuk** [comment]
> Also we have a warning on the None `is_delete_operator_pod`, what is the policy for the providers deprecation?

We don't have provider deprecation policy - we remove deprecated things when we see that they woudl be helpful in cleaning up the code.

**Comment by potiuk** [comment]
BTW. @eladkal is big proponent of removal of all deprecations when we release a breaking version. I am a bit more cautiouss, but I have no problem with that.

**Comment by hussein-awala** [comment]
IMHO we can merge this PR without updating the operators `GKEStartPodOperator` and `EksPodOperator`, and I can create two PRs which do the same thing for these two operators and change the default value for pod_deletion to be consistent with KPO, and we can add them to the next major release of each provider with a breaking change note. WDYT?

**Comment by eladkal** [comment]
> IMHO we can merge this PR without updating the operators GKEStartPodOperator and EksPodOperator, and I can create two PRs which do the same thing for these two operators and change the default value for pod_deletion to be consistent with KPO, and we can add them to the next major release of each provider with a breaking change note. WDYT?

It's OK to update multiple providers in one PR.

I see no reason to break backward compatibility in any provider.
This PR add new parameter and deprecate the old one.
We should do the same for the other providers that inherit from KPO.

**Comment by hussein-awala** [comment]
Okay, I'll add them.

**Comment by hussein-awala** [comment]
Since `on_finish_action` is a new parameter in the KPO class, we cannot deprecate the old parameter in the subclasses without checking if the `cncf-kubernetes` provider version supports the new parameter. To accomplish that, I had two options:

Checking the provider version:
```python
if sys.version_info < (3, 8):
    from importlib_metadata import version
else:
    from importlib.metadata import version

kubernetes_provider_version = version("apache-airflow-providers-cncf-kubernetes")
if kubernetes_provider_version < "7.1.0":
  ...
```
However, this approach doesn't work in development environments that use the code from source without installing the provider lib.

The second option is to check if the KPO init method has the `on_finish_action` parameter (currently being used):
```python
kpo_init_args = inspect.signature(KubernetesPodOperator.__init__).parameters

if "on_finish_action" not in kpo_init_args:
  ...
```
WDYT?

**Comment by eladkal** [comment]
>  Since on_finish_action is a new parameter in the KPO class, we cannot deprecate the old parameter in the subclasses without checking if the cncf-kubernetes provider version supports the new parameter

I don't think this is an issue.

Users who use the new versions of cncf.kubernetes will get the warning wether they use KPO or subclass of KPO.
So as long as the warning is there users will get notified about it.

**Comment by hussein-awala** [comment]
The issue with these operators is that the default `is_delete_operator_pod` is not aligned with the KPO. Therefore, we need to handle this parameter individually for each subclass operator.

Using `on_finish_action` as an argument for the super function won't work with the older version of the KPO. Additionally, providing `is_delete_operator_pod` will result in a deprecation warning being logged, even if the user provides `on_finish_action` for the GKE/EKS operator.
IMO, the best solution is to check the KPO version until modifying the default value of `is_delete_operator_pod` for the GKE/EKS operator, which, btw, is already deprecated.

**Comment by kand617** [comment]
Any update on this?
Our usecase is to keep pods that have failed and delete ones that completed successfully. 
Can we opt for a simple solution? A new param perhaps? 
 `is_delete_operator_pod_failed` `is_delete_operator_pod_success`

if `is_delete_operator_pod=True`  without extra params => delete pod always. 
if either new params are defined, then ignore is_delete_operator_pod?

**Comment by eladkal** [comment]
> Using `on_finish_action` as an argument for the super function won't work with the older version of the KPO. Additionally, providing `is_delete_operator_pod` will result in a deprecation warning being logged, even if the user provides `on_finish_action` for the GKE/EKS operator.


Is this a concern only for google provider?
if so I think we can just bump the k8s provider version dependency for google.
I'm suggesting it because currently google provider depended on  
https://github.com/apache/airflow/blob/dc5bf3fd02c589578209cb0dd5b7d005b1516ae9/airflow/providers/google/provider.yaml#L1121-L1123
 which means it must use the latest version which is 7.0.0
 I think bumping the dependency as `apache-airflow-providers-cncf-kubernetes>7.0.0` the difference between 7.0.0 to 7.1.0 will be very little and it will solve your concern.

**Comment by hussein-awala** [comment]
> Should we consider bumping the dependency apache-airflow-providers-cncf-kubernetes>7.0.0 ?

@eladkal +1, bu we should add the same dependency to AWS provider, is that possible?

**Review by eladkal — APPROVED** [review]

File: airflow/providers/amazon/provider.yaml
Comment (eladkal) [inline airflow/providers/amazon/provider.yaml]: I don't think 8.0.0 is right?
There is no such version for cncf.kubernetes

I believe it should be:
`apache-airflow-providers-cncf-kubernetes>7.0.0`

Comment (potiuk) [reply airflow/providers/amazon/provider.yaml]: I think `>=8.0.0` is fine, and we have to release those three packages together. There should be 8.0.0 release of cncf.kubernetes coming together with releasing of those two providers (the changes in cncf.kubernetes it brings are breaking, so that's expected to be bumped to 8.0.0.

Comment: The "additional" dependencies are pretty special because they are mostly "informatory". Almost no-one will install `apache-airflow-providers-amazon[cncf.kubernetes]==8.3.0` - what is mainly the purpose of those is to be document the dependencies.

**Review by potiuk — APPROVED** [review]

File: airflow/providers/amazon/provider.yaml
Comment (eladkal) [reply airflow/providers/amazon/provider.yaml]: w8.. I'm missing something here.
In this PR we add a feature to `apache-airflow-providers-cncf-kubernetes` that feature is what needed for amazon and google provider so what we are after is setting dependency for the next version of `apache-airflow-providers-cncf-kubernetes` thus the condition should be:

`apache-airflow-providers-cncf-kubernetes>=7.2.0`

Currently stable version of cncf.kubernetes is 7.1.0

Comment (potiuk) [reply airflow/providers/amazon/provider.yaml]: > In this PR we add a feature to `apache-airflow-providers-cncf-kubernetes` that feature is what needed for amazon and google provider so what we are after is setting dependency for the next version of `apache-airflow-providers-cncf-kubernetes` thus the condition should be:

Ah ok. If that's not a breaking change, then yes. I have not looked in detail enough and badly classified upcoming cncf.kubernetes version as MAJOR upgrade. 

This is really the result of our release process and why in the past I was leaving all the version upgrades to the release manager "documentation preparation" - because literally we do not know if the next version will be MINOR or MAJOR upgrade until we merge last change. I think it's next to impossible to change it unless we would have a much more active RM role upfront (RM would have to make versioning decisions upfront and for example hold breaking changes from being merged if the next release is non-breaking, but I feel that is quite impossible with our volunteer-driven organisation). 

Ideally we should have in this case `>= NEXT_CNCF_KUBERNETES_PROVIDER_VERSION` that could be automatically changed when the release documentation is prepared - because this is what we have here, really. But it's yet another automation complexity and feels like a bit of an edge-case that can be handled manually. 

And in edge-cases like that it's hard to have a process to follow for every single case like that, so it's case-by-case basis. 

Side note. Assuming that we do not know if the next release is MINOR or MAJOR we could add `>=MINOR_UPGRADE` always. It's a bit limiting and slightly confusing for the users if we do MAJOR release instead, but technically it would work.

For example in this case we could agree to have `>=7.2.0` here independently if the next version is MAJOR or MINOR. Even if we release 8.0.0, that would still technicaly work (8.0.0 >= 7.2.0), but there would be no 7.2.0 release at all., However (except confusing the users) there is also a side effect that it would not alow us (we don't do it anyway but techniclaly there is such a possiblity with mixed-governance model) to release 7.2.0 after the 8.0.0. Technically (and purely theoretically because we've never done so) we **could** have it and then the `>=7.2.0` would be wrong if the features needed will not be cherry-picked to the v7 branch of the provider.

But I think it's better to get it **right** even if it requires a bit of discussion and "humans-in-the-loop" agreeing on details of such versioning question (as we do now).

**Comment by potiuk** [comment]
`>=7.2.0` @hussein-awala ?

File: airflow/providers/amazon/provider.yaml
Comment (potiuk) [inline airflow/providers/amazon/provider.yaml]:
```suggestion
      - apache-airflow-providers-cncf-kubernetes>=7.2.0
```

File: airflow/providers/google/provider.yaml
Comment (potiuk) [inline airflow/providers/google/provider.yaml]:
```suggestion
      - apache-airflow-providers-cncf-kubernetes>=7.2.0
```

File: airflow/providers/amazon/provider.yaml
Comment (hussein-awala) [reply airflow/providers/amazon/provider.yaml]: > Side note. Assuming that we do not know if the next release is MINOR or MAJOR we could add >=MINOR_UPGRADE always. It's a bit limiting and slightly confusing for the users if we do MAJOR release instead, but technically it would work.

I like this idea, I added 8.0.0 because I didn't know if there is any breaking changes on the other PRs, but yes 7.2.0 will work for both minor and major releases 👍

**Review by eladkal — APPROVED** [review]

**Comment by eladkal** [comment]
Great work @hussein-awala !

**Pull request merged**
